backend/drm/backend.c

```diff
@@ -0,0 +1,118 @@
+#include <unistd.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include <errno.h>
+
+#include <wlr/common/list.h>
+
+#include "backend/drm/backend.h"
+#include "backend/drm/drm.h"
+#include "backend/drm/session.h"
+#include "backend/drm/udev.h"
+#include "common/log.h"
+
+struct wlr_drm_backend *wlr_drm_backend_init(struct wl_listener *add,
+		struct wl_listener *rem,
+		struct wl_listener *render)
+{
+	struct wlr_drm_backend *backend = calloc(1, sizeof *backend);
+	if (!backend) {
+		wlr_log(L_ERROR, "Allocation failed: %s", strerror(errno));
+		return NULL;
+	}
+
+	backend->displays = list_create();
+	if (!backend->displays) {
+		wlr_log(L_ERROR, "Failed to allocate list");
+		goto error_backend;
+	}
+
+	backend->event_loop = wl_event_loop_create();
+	if (!backend->event_loop) {
+		wlr_log(L_ERROR, "Failed to create event loop");
+		goto error_list;
+	}
+
+	if (!wlr_session_start(&backend->session)) {
+		wlr_log(L_ERROR, "Failed to start session");
+		goto error_loop;
+	}
+
+	if (!wlr_udev_init(backend)) {
+		wlr_log(L_ERROR, "Failed to start udev");
+		goto error_session;
+	}
+
+	backend->fd = wlr_udev_find_gpu(&backend->udev, &backend->session);
+	if (backend->fd == -1) {
+		wlr_log(L_ERROR, "Failed to open DRM device");
+		goto error_udev;
+	}
+
+	if (!wlr_drm_renderer_init(&backend->renderer, backend, backend->fd)) {
```

Comment:
At some point I would like to detach this somewhat from the backend as well, but for now it's fine.